### PR TITLE
Complex genomes

### PR DESCRIPTION
**Pull request by KatrinaLiqy**
Updates for modularity/diversity functions, testing, and datafile generation

### CODE REVIEW

**Comment by KatrinaLiqy** [comment]
resolved merge conflicts

File: SymSettings.cfg [hunk 1]
```diff
@@ -12,7 +12,7 @@ set GRID_Y 100                    # Height of world, just multiplied by width to
 set POP_SIZE -1                   # Starting size of the host population, -1 for full starting population
 set SYM_LIMIT 1                   # Number of symbiont allowed to infect a single host
 set START_MOI 1                   # Ratio of symbionts to hosts that experiment should start with
-set UPDATES 1001                  # Number of updates to run before quitting
+set UPDATES 2000                  # Number of updates to run before quitting
```
Comment (anyaevostinar) [inline SymSettings.cfg:15]: Make sure not to add SymSettings so that the file in the repository isn't changed. You should revert this change in your fork

File: source/native/symbulation_sgp.cc [hunk 2]
```diff
@@ -26,48 +27,67 @@ int symbulation_main(int argc, char *argv[]) {
...
+    task_set = SquareTasks;
+  } else if (config.TASK_TYPE() == 1) {
     task_set = LogicTasks;
   }
-
   SGPWorld world(random, &config, task_set);
 
   worldSetup(&world, &config);
   world.CreateDataFiles();
+  // Print some debug info for testing purposes
```
Comment (anyaevostinar) [inline source/native/symbulation_sgp.cc:39]: You should remove this comment and the commented out code below

File: source/native/symbulation_sgp.cc [hunk 3]
```diff
@@ -26,48 +27,67 @@ int symbulation_main(int argc, char *argv[]) {
...
+  //}
+  world.OnAnalyzePopulation([&](){
+    emp::vector<CPU> cpus = {};
+    emp::vector<std::pair<emp::Ptr<Organism>, size_t>> dominant_organisms =
+      world.GetDominantInfo();
+    for (auto pair : dominant_organisms) {
+      auto sample = pair.first.DynamicCast<SGPHost>();
+      cpus.push_back(sample->GetCPU());
+      if(sample->HasSym()){
+        cpus.push_back(sample->GetSymbionts().front().DynamicCast<SGPSymbiont>()->GetCPU());
```
Comment (anyaevostinar) [inline source/native/symbulation_sgp.cc:53]: I don't think you should have the hosts and symbionts mixed together for diversity analysis, you should get the diversity metrics of each population seperately

File: source/native/symbulation_sgp.cc [hunk 4]
```diff
@@ -26,48 +27,67 @@ int symbulation_main(int argc, char *argv[]) {
...
             << std::endl;
 
-  // Print some debug info for testing purposes
-  std::string file_ending = "_SEED" + std::to_string(config.SEED()) + ".data";
   {
-    size_t idx = 0;
-    for (auto pair : dominant_organisms) {
-      auto sample = pair.first.DynamicCast<SGPHost>();
-
+    ofstream genome_file;
```
Comment (anyaevostinar) [inline source/native/symbulation_sgp.cc:77]: This is the older version of this code, it looks like you accepted the incorrect version when resolving the conflicts, you should grab the version that is on my branch currently.

File: source/sgp_mode/DiversityAnalysis.h [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+#include "sgpl/operations/unary/Increment.hpp"
+#include "sgpl/operations/unary/Terminal.hpp"
+#include "sgpl/program/Program.hpp"
+#include "sgpl/spec/Spec.hpp"
+#include "sgpl/utility/ThreadLocalRandom.hpp"
+#include <math.h>
+#include <set>
+
+
+emp::BitSet<64> ReturnTasksDone(CPU org_cpu) {
```
Comment (anyaevostinar) [inline source/sgp_mode/DiversityAnalysis.h:34]: You should add documentation for each of these functions

File: source/sgp_mode/DiversityAnalysis.h [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+  org_cpu.RunCPUStep(emp::WorldPosition::invalid_id, 100);
+  
+
+  return *org_cpu.state.used_resources;
+}
+
+
+
+
+emp::unordered_map<emp::BitSet<64>, int> GetPhenotypeMap(emp::vector<CPU> organisms){ // SGPWorld world
```
Comment (anyaevostinar) [inline source/sgp_mode/DiversityAnalysis.h:45]: Please clean up these comments since it doesn't look like they are informative

File: source/sgp_mode/DiversityAnalysis.h [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+    }
+    return phenotype;
+
+
+}
+
+
+
+double AlphaDiversityHelper(emp::unordered_map<emp::BitSet<64> , int> phenotype, double q = 0.0) {
+    int pop = 0;
```
Comment (anyaevostinar) [inline source/sgp_mode/DiversityAnalysis.h:64]: let's make this a more descriptive variable name since 'pop' is already used to hold the entire population in World and it isn't clear what this is doing

File: source/sgp_mode/DiversityAnalysis.h [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+
+  return *org_cpu.state.used_resources;
+}
+
+
+
+
+emp::unordered_map<emp::BitSet<64>, int> GetPhenotypeMap(emp::vector<CPU> organisms){ // SGPWorld world
+//pointer??
+    emp::unordered_map<emp::BitSet<64> , int> phenotype;
```
Comment (anyaevostinar) [inline source/sgp_mode/DiversityAnalysis.h:47]: Through out this file, variable names could be improved to make it easier to understand what the code is doing. For example, 'phenotype_counts' would be more clear than phenotype

File: source/sgp_mode/SGPWorld.h [hunk 9]
```diff
@@ -44,6 +44,8 @@ class SGPWorld : public SymWorld {
   emp::Ptr<SyncDataMonitor<double>> data_node_sym_donated;
   emp::Ptr<SyncDataMonitor<double>> data_node_sym_stolen;
   emp::Ptr<SyncDataMonitor<double>> data_node_sym_earned;
+  // emp::Ptr<SyncDataMonitor<double>> data_node_diversity;
```
Comment (anyaevostinar) [inline source/sgp_mode/SGPWorld.h:47]: Why are these commented out?

File: source/test/sgp_mode_test/SGPDiversity.test.cc [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+    int pop_size = 1;
+    TaskSet test_tasks = LogicTasks;
+    SGPWorld world(random, &config, test_tasks);
+    world.resize(world_size);
+    config.POP_SIZE(pop_size);
+
+    size_t length = 100;
+
+    ProgramBuilder builder;
+    // builder.AddNand();
```
Comment (anyaevostinar) [inline source/test/sgp_mode_test/SGPDiversity.test.cc:39]: Why are these commented out?

File: source/test/sgp_mode_test/SGPDiversity.test.cc [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+    expected_map.insert({zeros,2});
+    emp::vector<CPU> test_vec = {};
+    test_vec.push_back((*test_sample1).GetCPU());
+    test_vec.push_back((*test_sample2).GetCPU());
+
+    
+    
+    emp::unordered_map<emp::BitSet<64>, int> test_map = GetPhenotypeMap(test_vec);
+
+    THEN("") { REQUIRE(test_map[zeros] == expected_map[zeros]); }
```
Comment (anyaevostinar) [inline source/test/sgp_mode_test/SGPDiversity.test.cc:62]: You should have something in the THEN quotes to describe the result

File: source/test/sgp_mode_test/SGPDiversity.test.cc [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+    int pop_size = 1;
+    TaskSet test_tasks = LogicTasks;
+    SGPWorld world(random, &config, test_tasks);
+    world.resize(world_size);
+    config.POP_SIZE(pop_size);
+
+    size_t length = 100;
+
+    ProgramBuilder builder;
+    // builder.AddNand();
```
Comment (anyaevostinar) [inline source/test/sgp_mode_test/SGPDiversity.test.cc:85]: Again remove the commented out code

File: source/test/sgp_mode_test/SGPDiversity.test.cc [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+    expected_map.insert({zeros,2});
+    double expected_value = 1.0; 
+    emp::vector<CPU> test_vec = {};
+    test_vec.push_back((*test_sample1).GetCPU());
+    test_vec.push_back((*test_sample2).GetCPU());
+
+    
+    double test_value = AlphaDiversity(test_vec);
+
+    THEN("") { REQUIRE(test_value == expected_value); }
```
Comment (anyaevostinar) [inline source/test/sgp_mode_test/SGPDiversity.test.cc:108]: Again the THEN

**Review by anyaevostinar — CHANGES_REQUESTED** [review]
Some changes needed

File: SymSettings.cfg (hunk 1 above)
Comment (anyaevostinar) [reply SymSettings.cfg:15]: This still needs to be fixed

Comment (anyaevostinar) [inline SymSettings.cfg:15]: This still needs to be set back to 1001

File: source/native/symbulation_sgp.cc [hunk 14]
```diff
@@ -36,33 +36,38 @@ int symbulation_main(int argc, char *argv[]) {
...
       auto sample = pair.first.DynamicCast<SGPHost>();
-      cpus.push_back(sample->GetCPU());
+      host_cpus.push_back(sample->GetCPU());
+
       if(sample->HasSym()){
-        cpus.push_back(sample->GetSymbionts().front().DynamicCast<SGPSymbiont>()->GetCPU());
+        for (auto sym: sample->GetSymbionts()){
+          sym_cpus.push_back(sym.DynamicCast<SGPSymbiont>()->GetCPU());
+        }
+        // cpus.push_back(sample->GetSymbionts().front().DynamicCast<SGPSymbiont>()->GetCPU());
```
Comment (anyaevostinar) [inline source/native/symbulation_sgp.cc:55]: Remember to remove commented out code

File: source/sgp_mode/DiversityAnalysis.h [hunk 15]
```diff
@@ -30,85 +30,182 @@
...
+ *
+ * Input: A map containing the count of phenotypes
+ *
+ * Output: The alpha diversity of the population
+ *
+ * Purpose: Calculates the alpha diversity based on the formula
+ *
+ */
+double AlphaDiversityHelper(emp::vector<emp::BitSet<64>> phenotypes, int j = 4, double q = 0.0) {
+    // if (q == 0.0){
```
Comment (anyaevostinar) [inline source/sgp_mode/DiversityAnalysis.h:108]: Remember to remove commented out debugging code

File: source/sgp_mode/DiversityAnalysis.h [hunk 16]
```diff
@@ -30,85 +30,182 @@
...
+    //     return 1.0;
+    // }
+    // std::cout<< "break point2" << std::endl;
+    emp::Random random = emp::Random();
```
Comment (anyaevostinar) [inline source/sgp_mode/DiversityAnalysis.h:112]: You shouldn't be using a new random, instead you should use the existing one or at least make sure this is seeded properly so that it is reproducible. Also, why do you need random numbers for this analysis?

File: source/sgp_mode/DiversityAnalysis.h [hunk 17]
```diff
@@ -30,85 +30,182 @@
...
+/**
+ *
+ * Input: A map containing the count of phenotypes
+ *
+ * Output: The alpha diversity of the population
+ *
+ * Purpose: Calculates the alpha diversity based on the formula
+ *
+ */
+double AlphaDiversityHelper(emp::vector<emp::BitSet<64>> phenotypes, int j = 4, double q = 0.0) {
```
Comment (anyaevostinar) [inline source/sgp_mode/DiversityAnalysis.h:107]: You should include the optional parameters j and q in the documentation for what they are

File: source/sgp_mode/ModularityAnalysis.h [hunk 18]
```diff
@@ -107,6 +107,7 @@ emp::vector<int> GetUsefulEnds(emp::vector<emp::vector<int>> task_programs) {
 int GetNumSites(int start_inst, int end_inst, emp::vector<int> alt_genome) {
   // for altered genome clusters
   int total_sites = 0;
+  // int genome_size = (int)alt_genome.size();
```
Comment (anyaevostinar) [inline source/sgp_mode/ModularityAnalysis.h:110]: Remove commented out code

File: source/sgp_mode/ModularityAnalysis.h [hunk 19]
```diff
@@ -191,7 +192,7 @@ double GetPModularity(emp::vector<emp::vector<int>> task_programs) {
 
 /**
  *
- * Input:Takes in the number of tasks in the taskset and a CPU reference
+ * Input:Takes in the number of tasks in the taskset and a pointer to a host
```
Comment (anyaevostinar) [inline source/sgp_mode/ModularityAnalysis.h:195]: Doesn't look like this is accurate

File: source/sgp_mode/ModularityAnalysis.h [hunk 20]
```diff
@@ -261,7 +262,7 @@ double GetFModularity(emp::vector<emp::vector<int>> task_programs) {
...
```
Comment (anyaevostinar) [inline source/sgp_mode/ModularityAnalysis.h:265]: Also not accurate

File: source/test/sgp_mode_test/SGPDiversity.test.cc [hunk 21]
```diff
@@ -105,7 +104,9 @@ TEST_CASE("AlphaDiversity", "[sgp]") {
     
     double test_value = AlphaDiversity(test_vec);
 
-    THEN("") { REQUIRE(test_value == expected_value); }
+    THEN("Calculation is correct") {
```
Comment (anyaevostinar) [inline source/test/sgp_mode_test/SGPDiversity.test.cc:107]: Instead of just 'calculation is correct', you should state what it should actually be: "alpha diversity is equal to 1"

File: source/native/symbulation_sgp.cc (hunk 4 above)
Comment (anyaevostinar) [reply source/native/symbulation_sgp.cc:82]: This still needs to be fixed

**Review by anyaevostinar — CHANGES_REQUESTED** [review]
Some things still need to be fixed, be sure to look through and address all the comments

File: source/native/symbulation_sgp.cc [hunk 22]
```diff
@@ -71,28 +120,30 @@ int symbulation_main(int argc, char *argv[]) {
...
-                   *sample,
-                   *sample->GetSymbionts().front().DynamicCast<SGPSymbiont>(),
-                   world)
-            << std::endl;
-      } else {
-        // We want something in the file so it overwrites previous data, and NA
-        // is something R generally understands
-        mutualism_file << "NA" << std::endl;
-      }
+  ofstream modularity_file;
```
Comment (anyaevostinar) [inline source/native/symbulation_sgp.cc:123]: It looks like you got rid of the symbiont impact file code, we need both of those data files

File: source/sgp_mode/AnalysisTools.h [hunk 23]
```diff
@@ -32,40 +32,73 @@
 
 // start of getNecessarySites methods
 
-/*
- *
- * Input: Takes in a cpu and the identifier for a specific task
+using Library = sgpl::OpLibrary<
```
Comment (anyaevostinar) [inline source/sgp_mode/AnalysisTools.h:35]: This is duplicated from code in GenomeLibrary.h and should be removed, remember to be careful to retain the correct versions of the code when doing merges

File: source/sgp_mode/AnalysisTools.h [hunk 24]
```diff
@@ -32,40 +32,73 @@
...
+ * Input:
  *
- * Output: Returns a vector representing the full genome, reduced to 1s and 0s to show either
- *that an instruction is necessary to complete the task, or not respectively
+ * Output:
  *
- *Purpose: Is to return a vector that acts as a reduced program representation
- *of the necessary code lines to complete the given task
+ *Purpose: Takes all the calculation methods and calls them in order of having a
+ *simplified way of getting an organism's Physical Modularity
```
Comment (anyaevostinar) [inline source/sgp_mode/AnalysisTools.h:101]: This documentation looks like it got put in the wrong place somehow, doesn't apply to the code below and removes documentation that did look correct

File: source/sgp_mode/AnalysisTools.h [hunk 25]
```diff
@@ -85,7 +118,7 @@ emp::vector<int> GetNecessaryInstructions(CPU org_cpu, size_t test_task_id) {
 
   if (can_do_task) {
 
-    for (size_t k = 0; k <= control_program.size() - 1; k++) {
+    for (int k = 0; k <= control_program.size() - 1; k++) {
```
Comment (anyaevostinar) [inline source/sgp_mode/AnalysisTools.h:121]: size_t is the more appropriate type, so you should go with the previous version of the code

File: source/sgp_mode/AnalysisTools.h [hunk 26]
```diff
@@ -109,15 +142,12 @@ emp::vector<int> GetNecessaryInstructions(CPU org_cpu, size_t test_task_id) {
 
 /**
  *
- * Input: Takes in an organism's CPU
+ * Input:
```
Comment (anyaevostinar) [inline source/sgp_mode/AnalysisTools.h:145]: The previous version of this documentation looks more complete

File: source/sgp_mode/DiversityAnalysis.h [hunk 27]
```diff
@@ -60,6 +44,7 @@ emp::BitSet<64> ReturnTasksDone(CPU org_cpu) {
  */
 emp::unordered_map<emp::BitSet<64>, int> GetPhenotypeMap(emp::vector<CPU> organisms){
     emp::unordered_map<emp::BitSet<64> , int> phenotype_counts;
+    std::cout << "size: " << organisms.size()<< std::endl;
```
Comment (anyaevostinar) [inline source/sgp_mode/DiversityAnalysis.h:47]: Remember to remove debugging print statements before you submit to a pull request

File: source/sgp_mode/DiversityAnalysis.h [hunk 28]
```diff
@@ -97,31 +82,47 @@ emp::vector<emp::BitSet<64>> GetPhenotypeVector(emp::vector<CPU> organisms){
...
-    // if (q == 0.0){
-    //     return 1.0;
-    // }
-    // std::cout<< "break point2" << std::endl;
-    emp::Random random = emp::Random();
+double AlphaDiversityHelper(emp::Random random, emp::vector<emp::BitSet<64>> phenotypes, int j = 9, double q = 0.0) {
+    
+    
     int org_count = phenotypes.size();
+    std::cout << "total size: " << org_count << std::endl;
```
Comment (anyaevostinar) [inline source/sgp_mode/DiversityAnalysis.h:116]: Print statement to remove

File: source/sgp_mode/DiversityAnalysis.h [hunk 29]
```diff
@@ -97,31 +82,47 @@ emp::vector<emp::BitSet<64>> GetPhenotypeVector(emp::vector<CPU> organisms){
...
+        std::cout << "phenotype: " << phenotype.first << " count: " << phenotype.second<< std::endl;
+        i += 1;
+    }
+    std::cout << "richness: " << i << std::endl;
+    return i;
+}
+
+/**
+ *
+ * Input: A random object (of the world), A map containing the count of phenotypes
```
Comment (anyaevostinar) [inline source/sgp_mode/DiversityAnalysis.h:105]: You still need to add documentation for the two optional parameters

File: source/test/sgp_mode_test/SGPDiversity.test.cc (hunk 21 above)
Comment (anyaevostinar) [reply source/test/sgp_mode_test/SGPDiversity.test.cc:107]: This still needs to be fixed

File: source/test/sgp_mode_test/SGPDiversity.test.cc [hunk 30]
```diff
@@ -0,0 +1,162 @@
...
+#include "../../sgp_mode/SGPDataNodes.h"
+#include "../../sgp_mode/SGPHost.h"
+#include "../../sgp_mode/SGPSymbiont.h"
+#include "../../sgp_mode/SGPWorld.h"
+#include "../../sgp_mode/Scheduler.h"
+#include "../../sgp_mode/Tasks.h"
+#include <iostream>
+#include <memory>
+
+
```
Comment (anyaevostinar) [inline source/test/sgp_mode_test/SGPDiversity.test.cc:22]: You should definitely add a test for species richness as well

**Review by anyaevostinar — CHANGES_REQUESTED** [review]
More things to fix and some older comments that haven't been addressed yet

**Review by anyaevostinar — APPROVED** [review]
Looks good
